### ISSUE TITLE
Unjustified typescript error when using for..of on a NodeList

### ISSUE THREAD

**Issue by thany** [microsoft/TypeScript#31656]
Issue Type: <b>Bug</b>

Consider this code:

```typescript
const elements = document.querySelectorAll('div');
for (const element of elements) {
    console.log(element);
}
```

The editor will produce an "errror" on `elements` saying:

> Type 'NodeListOf<Element>' is not an array type or a string type. (ts2495)

While that message is technically correct, there isn't a reason to produce an error there in the first place, because a NodeList is perfectly iterable. The same problem happens when spreading a NodeList into an array, like so: `[...elements]`.

I don't understand why typescript is complaining about this **ONLY in the editor**. The code compiles perfectly fine, so there's not anything I have to do in my compiler settings. Not only does it compile, it even *works*.

In the editor, VSCode seems to be using typescript 3.4.5. Surely by now, typescript knows that there are more iterable objects than only arrays and strings. As far as I can tell, there's something wrong with the built-in typescript engine.

This problem has existed for as long as I'm using typescript, which is a few months, and apparently this problem made it unfixed in 1.34, which I'm sad to find out about.

VS Code version: Code 1.34.0 (a622c65b2c713c890fcf4fbf07cf34049d5fe758, 2019-05-15T21:59:37.030Z)
OS version: Windows_NT x64 10.0.17763

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz (12 x 2904)|
|GPU Status|2d_canvas: enabled<br>checker_imaging: disabled_off<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>native_gpu_memory_buffers: disabled_software<br>rasterization: enabled<br>surface_synchronization: enabled_on<br>video_decode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|31.74GB (16.20GB free)|
|Process Argv||
|Screen Reader|no|
|VM|67%|
</details><details><summary>Extensions (19)</summary>

Extension|Author (truncated)|Version
---|---|---
better-comments|aar|2.0.5
overtype|ada|0.2.0
Bookmarks|ale|10.4.3
color-info|bie|0.5.1
mustache|daw|1.1.1
vscode-eslint|dba|1.9.0
vscode-babel-coloring|dza|0.0.4
EditorConfig|Edi|0.13.0
vscode-great-icons|emm|2.1.47
prettier-vscode|esb|1.9.0
file-icons|fil|1.0.17
duplicate|gee|1.0.1
docthis|joe|0.7.1
vscode-duplicate|mrm|1.2.1
vscode-docker|Pet|0.6.2
vscode-versionlens|pfl|0.23.0
vscode-sort-json|ric|1.13.0
code-settings-sync|Sha|3.2.9
vscode-svg-previewer|vit|0.5.0

(1 theme extensions excluded)

</details>

**Comment by mjbvz** [comment]
This is because you have ` "target": "es5"`. I believe this is by design but someone on the TS team would have to clarify why. 

This config fixes the issue:

```json
"compilerOptions": {
    "module": "commonjs",
    "target": "es5",
    "downlevelIteration": true,
    "lib": [
        "es2016",
        "dom",
        "dom.iterable"
    ]
}
```

**Comment by thany** [comment]
Where should I put this config?

**Comment by heuperman** [comment]
You've probably figured it out by now but just in case: you should put that in the `tsconfig.json` file in your main project directory.

**Comment by thany** [comment]
I hadn't, but let me repeat that VScode is *not* using my project's typescript engine. It's using its own built-in engine. So why (and how) would it know to use my particular tsconfig?

**Comment by MrDesjardins** [comment]
@thany, you can click at the bottom of the VsCode editor the version number of TypeScript. It will open a dialog where you can select "Use workspace version". The workspace version is the `tsconfig.json` that is at the root of your project (the folder that contains all the source files of your application).

![image](https://user-images.githubusercontent.com/1014796/61104076-be9bb280-a429-11e9-8ce8-1ccf53c1354b.png)

**Comment by thany** [comment]
I don't have that option, for some reason.

Probably a bug in vscode if it *does* manage to find our tsconfig but *doesn't* know which typescript version it's for 🤷‍♂️
